File: apps/api/src/analytics/features.ts
import { Effect, Schema } from 'effect';
import { Infrastructure } from '../platform/resources';
import { attempt, invalid, ApiError, attemptSync } from '../shared/errors';
import { getEnvironment } from '../sites/service';
import { dateRange } from './reports';
import { overview } from './overview';
import { decode, Id, Name, id } from '../shared/validation';
import native from './sql/native.sql' with { type: 'text' };
const missing = () =>
  new ApiError({ status: 404, code: 'not_found', message: 'Feature not found.' });
export const configureFeature = Effect.fn('configureFeature')(function* (
  owner: string,
  site: string,
  key: string,
  feature: string,
  body: unknown,
) {
  const r = yield* Infrastructure;
  const env = yield* getEnvironment(owner, site, key);
  return yield* attempt(() =>
    r.primary.begin(async (tx) => {
      await tx`SELECT id FROM environments WHERE id=${env.id}::uuid FOR UPDATE`;
      if (feature === 'goals') {
        const input = decode(
          Schema.Struct({
            name: Name,
            matchType: Schema.Literals(['event', 'page']),
            matchValue: Schema.String.check(Schema.isMinLength(1)).check(Schema.isMaxLength(2048)),
          }),
          body,
        );
        if (
          input.matchType === 'event'
            ? !/^[a-zA-Z][a-zA-Z0-9_.-]{0,63}$/.test(input.matchValue)
            : !input.matchValue.startsWith('/') || /[?#]/.test(input.matchValue)
        )
          throw invalid('Use an event name or a page path without a query string.');
        const [count] =
          await tx`SELECT count(*)::int AS n FROM conversion_goals WHERE environment_id=${env.id}::uuid`;
        if (count.n >= 20) throw invalid('An environment can have up to 20 goals.');
        const [row] =
          await tx`INSERT INTO conversion_goals(environment_id,name,match_type,match_value) VALUES(${env.id}::uuid,${input.name.trim()},${input.matchType},${input.matchValue}) ON CONFLICT DO NOTHING RETURNING id,name,match_type AS "matchType",match_value AS "matchValue"`;
        if (!row) throw invalid('A goal already tracks this conversion.');
        return { goal: row };
      }
      if (feature === 'annotations') {
        const input = decode(
          Schema.Union([
            Schema.Struct({
              action: Schema.Literal('create'),
              day: Schema.String,
              label: Schema.String.check(Schema.isMinLength(1)).check(Schema.isMaxLength(120)),
            }),
            Schema.Struct({ action: Schema.Literal('delete'), id: Id }),
          ]),
          body,
        );
        if (input.action === 'delete') {
          const rows =
            await tx`DELETE FROM overview_annotations WHERE environment_id=${env.id}::uuid AND id=${input.id}::uuid RETURNING id`;
          if (!rows.length) throw missing();
          return { deleted: true };
        }
        const range = dateRange({ from: input.day, to: input.day });
        const [count] =
          await tx`SELECT count(*)::int AS n FROM overview_annotations WHERE environment_id=${env.id}::uuid`;
        if (count.n >= 500) throw invalid('An environment can have up to 500 annotations.');
        const [row] =
          await tx`INSERT INTO overview_annotations(environment_id,day,label) VALUES(${env.id}::uuid,${range.from},${input.label.trim()}) RETURNING id,day::text,label`;
        return { annotation: row };
      }
      if (feature === 'errors') {
        const input = decode(
          Schema.Struct({ fingerprint: Schema.String.check(Schema.isPattern(/^[a-f0-9]{64}$/i)) }),
          body,
        );
        const exists =
          await r.analytics`SELECT 1 FROM diagnostic_events WHERE environment_id=${env.id}::uuid AND fingerprint=${input.fingerprint} AND kind='error' LIMIT 1`;
        if (exists.length)
          await tx`INSERT INTO error_resolutions(environment_id,fingerprint) VALUES(${env.id}::uuid,${input.fingerprint}) ON CONFLICT(environment_id,fingerprint) DO UPDATE SET resolved_at=now()`;
        return { resolved: true };
      }
      throw missing();
    }),
  );
});
export const deleteGoal = Effect.fn('deleteGoal')(function* (
  owner: string,
  site: string,
  key: string,
  goal: string,
) {
  const r = yield* Infrastructure;
  const env = yield* getEnvironment(owner, site, key);
  const rows = yield* attempt(
    () =>
      r.primary`DELETE FROM conversion_goals WHERE environment_id=${env.id}::uuid AND id=${id(goal)}::uuid RETURNING id`,
  );
  if (!rows.length) return yield* missing();
  return { deleted: true };
});
export const readFeature = Effect.fn('readFeature')(function* (
  owner: string,
  site: string,
  key: string,
  feature: string,
  query: Record<string, string>,
) {
  const r = yield* Infrastructure;
  const env = yield* getEnvironment(owner, site, key),
    range = yield* attemptSync(() => dateRange(query));
  if (feature === 'overview') return yield* overview(owner, site, key, query);
  return yield* attempt(async () => {
    const rows = await r.analytics.begin(async (tx) => {
      await tx`SET TRANSACTION READ ONLY`;
      if (feature === 'live')
        return tx`WITH recent AS MATERIALIZED (SELECT * FROM events WHERE site_id=${env.id}::uuid AND day>=current_date-1 AND received_at>now()-interval '5 minutes') SELECT jsonb_build_object('active',(SELECT count(DISTINCT(day,visitor)) FROM recent),'recent',coalesce((SELECT jsonb_agg(jsonb_build_object('path',path,'country',country,'at',received_at) ORDER BY received_at DESC) FROM (SELECT * FROM recent ORDER BY received_at DESC,id LIMIT 10) r),'[]'::jsonb))::text AS data`;
      if (feature === 'goals')
        return tx`SELECT goal_id,count(*)::bigint AS conversions,count(DISTINCT(day,visitor))::bigint AS visitors FROM goal_conversions WHERE environment_id=${env.id}::uuid AND day BETWEEN ${range.from} AND ${range.to} GROUP BY goal_id`;
      if (feature === 'web-vitals')
        return tx`SELECT jsonb_build_object('name',payload->>'name','device',device,'samples',count(*),'p75',percentile_cont(0.75) WITHIN GROUP(ORDER BY (payload->>'value')::double precision))::text AS data FROM diagnostic_events WHERE environment_id=${env.id}::uuid AND kind='vital' AND received_at>=${range.from}::date AND received_at<${range.to}::date+interval '1 day' GROUP BY payload->>'name',device ORDER BY payload->>'name',device`;
      if (feature === 'errors')
        return tx`SELECT jsonb_build_object('fingerprint',fingerprint,'message',(array_agg(payload->>'message' ORDER BY received_at DESC))[1],'source',(array_agg(payload->>'source' ORDER BY received_at DESC))[1],'stack',(array_agg(payload->>'stack' ORDER BY received_at DESC))[1],'path',(array_agg(path ORDER BY received_at DESC))[1],'occurrences',count(*),'visitors',count(DISTINCT(received_at::date,visitor)),'lastSeen',max(received_at),'resolved',false)::text AS data FROM diagnostic_events WHERE environment_id=${env.id}::uuid AND kind='error' AND received_at>=${range.from}::date AND received_at<${range.to}::date+interval '1 day' GROUP BY fingerprint ORDER BY max(received_at) DESC LIMIT 100`;
      throw missing();
    });
    if (feature === 'live') return JSON.parse(rows[0].data);
    if (feature === 'goals') {
      const definitions =
        await r.primary`SELECT id,name,match_type AS "matchType",match_value AS "matchValue",created_at AS "createdAt" FROM conversion_goals WHERE environment_id=${env.id}::uuid ORDER BY created_at,id`;
      const totals = await r.analytics.begin(async (tx) => {
        await tx`SET TRANSACTION READ ONLY`;
        return tx.unsafe(
          `${native} SELECT coalesce(sum(visitors),0)::bigint AS visitors FROM native`,
          [env.id, range.from, range.to],
        );
      });
      return {
        goals: definitions.map((goal: { id: string }) => {
          const count = rows.find((row: { goal_id: string }) => row.goal_id === goal.id);
          return {
            ...goal,
            conversions: Number(count?.conversions ?? 0),
            visitors: Number(count?.visitors ?? 0),
          };
        }),
        visitors: Number(totals[0].visitors),
      };
    }
    const items = rows.map((row: { data: string }) => JSON.parse(row.data));
    if (feature === 'errors') {
      const resolved =
        await r.primary`SELECT fingerprint,resolved_at FROM error_resolutions WHERE environment_id=${env.id}::uuid`;
      for (const item of items)
        item.resolved = resolved.some(
          (row: { fingerprint: string; resolved_at: Date }) =>
            row.fingerprint === item.fingerprint &&
            row.resolved_at.getTime() >= Date.parse(item.lastSeen),
        );
    }
    return { items };
  });
});

File: crates/api/src/analytics/features.rs
use super::{
    Parameters, overview, read_transaction,
    reports::{DateRange, NATIVE},
};
use crate::{AppState, error::ApiError, sites};
use axum::http::StatusCode;
use serde_json::{Value, json};
use sqlx::PgConnection;
use uuid::Uuid;

fn missing() -> ApiError {
    ApiError::new(StatusCode::NOT_FOUND, "not_found", "Feature not found.")
}

fn invalid(message: &'static str) -> ApiError {
    ApiError::new(StatusCode::BAD_REQUEST, "invalid_request", message)
}

fn fields(input: &Value, allowed: &[&str]) -> Result<(), ApiError> {
    if input
        .as_object()
        .is_none_or(|o| o.keys().any(|k| !allowed.contains(&k.as_str())))
    {
        return Err(ApiError::invalid());
    }
    Ok(())
}

pub(super) fn event_name(value: &str) -> bool {
    let mut chars = value.bytes();
    value.len() <= 64
        && chars.next().is_some_and(|c| c.is_ascii_alphabetic())
        && chars.all(|c| c.is_ascii_alphanumeric() || matches!(c, b'_' | b'.' | b'-'))
}

pub(super) async fn configure(
    state: &AppState,
    env: Uuid,
    feature: &str,
    input: Value,
) -> Result<Value, ApiError> {
    let mut tx = state.pool.begin().await?;
    sqlx::query("SELECT id FROM environments WHERE id=$1 FOR UPDATE")
        .bind(env)
        .execute(&mut *tx)
        .await?;
    let result = match feature {
        "goals" => {
            fields(&input, &["name", "matchType", "matchValue"])?;
            if !sites::valid_name(&input["name"]) {
                return Err(ApiError::invalid());
            }
            let name = input["name"].as_str().ok_or_else(ApiError::invalid)?.trim();
            let kind = input["matchType"].as_str().ok_or_else(ApiError::invalid)?;
            let value = input["matchValue"].as_str().ok_or_else(ApiError::invalid)?;
            if value.is_empty()
                || value.encode_utf16().count() > 2048
                || !matches!(kind, "page" | "event")
            {
                return Err(ApiError::invalid());
            }
            if if kind == "event" {
                !event_name(value)
            } else {
                !value.starts_with('/') || value.contains(['?', '#'])
            } {
                return Err(invalid(
                    "Use an event name or a page path without a query string.",
                ));
            }
            let count: i64 =
                sqlx::query_scalar("SELECT count(*) FROM conversion_goals WHERE environment_id=$1")
                    .bind(env)
                    .fetch_one(&mut *tx)
                    .await?;
            if count >= 20 {
                return Err(invalid("An environment can have up to 20 goals."));
            }
            let goal:Option<Value>=sqlx::query_scalar("INSERT INTO conversion_goals(environment_id,name,match_type,match_value) VALUES($1,$2,$3,$4) ON CONFLICT DO NOTHING RETURNING jsonb_build_object('id',id,'name',name,'matchType',match_type,'matchValue',match_value)")
                .bind(env).bind(name).bind(kind).bind(value).fetch_optional(&mut *tx).await?;
            json!({"goal":goal.ok_or_else(||invalid("A goal already tracks this conversion."))?})
        }
        "annotations" => match input["action"].as_str() {
            Some("delete") => {
                fields(&input, &["action", "id"])?;
                let id = sites::identifier(input["id"].as_str().ok_or_else(ApiError::invalid)?)?;
                let result = sqlx::query(
                    "DELETE FROM overview_annotations WHERE environment_id=$1 AND id=$2",
                )
                .bind(env)
                .bind(id)
                .execute(&mut *tx)
                .await?;
                if result.rows_affected() == 0 {
                    return Err(missing());
                }
                json!({"deleted":true})
            }
            Some("create") => {
                fields(&input, &["action", "day", "label"])?;
                let day = input["day"].as_str().ok_or_else(ApiError::invalid)?;
                let range = DateRange::parse(
                    &[("from".into(), day.into()), ("to".into(), day.into())].into(),
                )?;
                let label = input["label"]
                    .as_str()
                    .filter(|l| !l.is_empty() && l.encode_utf16().count() <= 120)
                    .ok_or_else(ApiError::invalid)?;
                let count: i64 = sqlx::query_scalar(
                    "SELECT count(*) FROM overview_annotations WHERE environment_id=$1",
                )
                .bind(env)
                .fetch_one(&mut *tx)
                .await?;
                if count >= 500 {
                    return Err(invalid("An environment can have up to 500 annotations."));
                }
                let row:Value=sqlx::query_scalar("INSERT INTO overview_annotations(environment_id,day,label) VALUES($1,$2,$3) RETURNING jsonb_build_object('id',id,'day',day::text,'label',label)")
                        .bind(env).bind(range.from).bind(label.trim()).fetch_one(&mut *tx).await?;
                json!({"annotation":row})
            }
            _ => return Err(ApiError::invalid()),
        },
        "errors" => {
            fields(&input, &["fingerprint"])?;
            let fingerprint = input["fingerprint"]
                .as_str()
                .filter(|s| s.len() == 64 && s.bytes().all(|b| b.is_ascii_hexdigit()))
                .ok_or_else(ApiError::invalid)?;
            sqlx::query("INSERT INTO error_resolutions(environment_id,fingerprint) SELECT $1,$2 WHERE EXISTS(SELECT 1 FROM diagnostic_events WHERE environment_id=$1 AND fingerprint=$2 AND kind='error') ON CONFLICT(environment_id,fingerprint) DO UPDATE SET resolved_at=now()")
                .bind(env).bind(fingerprint).execute(&mut *tx).await?;
            json!({"resolved":true})
        }
        _ => return Err(missing()),
    };
    tx.commit().await?;
    Ok(result)
}

pub(super) async fn delete_goal(
    state: &AppState,
    env: Uuid,
    goal: Uuid,
) -> Result<Value, ApiError> {
    if sqlx::query("DELETE FROM conversion_goals WHERE environment_id=$1 AND id=$2")
        .bind(env)
        .bind(goal)
        .execute(&state.pool)
        .await?
        .rows_affected()
        == 0
    {
        return Err(missing());
    }
    Ok(json!({"deleted":true}))
}

pub(super) async fn goals(
    connection: &mut PgConnection,
    env: Uuid,
    counts: Vec<Value>,
    visitors: i64,
) -> Result<Value, ApiError> {
    let mut definitions:Vec<Value>=sqlx::query_scalar("SELECT jsonb_build_object('id',id,'name',name,'matchType',match_type,'matchValue',match_value,'createdAt',created_at) FROM conversion_goals WHERE environment_id=$1 ORDER BY created_at,id")
        .bind(env).fetch_all(connection).await?;
    for goal in &mut definitions {
        let row = counts.iter().find(|c| c["goal_id"] == goal["id"]);
        goal["conversions"] = json!(row.and_then(|r| r["conversions"].as_i64()).unwrap_or(0));
        goal["visitors"] = json!(row.and_then(|r| r["visitors"].as_i64()).unwrap_or(0));
    }
    Ok(json!({"goals":definitions,"visitors":visitors}))
}

pub(super) async fn read(
    state: &AppState,
    env: Uuid,
    feature: &str,
    query: &Parameters,
) -> Result<Value, ApiError> {
    let range = DateRange::parse(query)?;
    if feature == "overview" {
        return overview::read(state, env, query).await;
    }
    let mut tx = read_transaction(state).await?;
    if feature == "live" {
        return Ok(sqlx::query_scalar("WITH recent AS MATERIALIZED (SELECT * FROM events WHERE site_id=$1 AND day>=current_date-1 AND received_at>now()-interval '5 minutes') SELECT jsonb_build_object('active',(SELECT count(DISTINCT(day,visitor)) FROM recent),'recent',coalesce((SELECT jsonb_agg(jsonb_build_object('path',path,'country',country,'at',received_at) ORDER BY received_at DESC) FROM (SELECT * FROM recent ORDER BY received_at DESC,id LIMIT 10) r),'[]'::jsonb))")
            .bind(env).fetch_one(&mut *tx).await?);
    }
    if feature == "goals" {
        let counts:Vec<Value>=sqlx::query_scalar("SELECT jsonb_build_object('goal_id',goal_id,'conversions',count(*),'visitors',count(DISTINCT(day,visitor))) FROM goal_conversions WHERE environment_id=$1 AND day BETWEEN $2 AND $3 GROUP BY goal_id")
            .bind(env).bind(range.from).bind(range.to).fetch_all(&mut *tx).await?;
        let visitors: i64 = sqlx::query_scalar(&format!(
            "{NATIVE} SELECT coalesce(sum(visitors),0)::bigint FROM native"
        ))
        .bind(env)
        .bind(range.from)
        .bind(range.to)
        .fetch_one(&mut *tx)
        .await?;
        return goals(&mut tx, env, counts, visitors).await;
    }
    let statement = match feature {
        "web-vitals" => {
            "SELECT jsonb_build_object('name',payload->>'name','device',device,'samples',count(*),'p75',percentile_cont(0.75) WITHIN GROUP(ORDER BY (payload->>'value')::double precision)) FROM diagnostic_events WHERE environment_id=$1 AND kind='vital' AND received_at>=$2::date AND received_at<$3::date+interval '1 day' GROUP BY payload->>'name',device ORDER BY payload->>'name',device"
        }
        "errors" => {
            "SELECT jsonb_build_object('fingerprint',fingerprint,'message',(array_agg(payload->>'message' ORDER BY received_at DESC))[1],'source',(array_agg(payload->>'source' ORDER BY received_at DESC))[1],'stack',(array_agg(payload->>'stack' ORDER BY received_at DESC))[1],'path',(array_agg(path ORDER BY received_at DESC))[1],'occurrences',count(*),'visitors',count(DISTINCT(received_at::date,visitor)),'lastSeen',max(received_at),'resolved',false) FROM diagnostic_events WHERE environment_id=$1 AND kind='error' AND received_at>=$2::date AND received_at<$3::date+interval '1 day' GROUP BY fingerprint ORDER BY max(received_at) DESC LIMIT 100"
        }
        _ => return Err(missing()),
    };
    let mut items: Vec<Value> = sqlx::query_scalar(statement)
        .bind(env)
        .bind(range.from)
        .bind(range.to)
        .fetch_all(&mut *tx)
        .await?;
    if feature == "errors" {
        let resolved: Vec<(String, chrono::DateTime<chrono::Utc>)> = sqlx::query_as(
            "SELECT fingerprint,resolved_at FROM error_resolutions WHERE environment_id=$1",
        )
        .bind(env)
        .fetch_all(&mut *tx)
        .await?;
        for item in &mut items {
            let last_seen = item["lastSeen"]
                .as_str()
                .and_then(|s| chrono::DateTime::parse_from_rfc3339(s).ok());
            item["resolved"] = json!(
                resolved
                    .iter()
                    .any(|(fingerprint, at)| item["fingerprint"] == *fingerprint
                        && last_seen.is_some_and(|seen| *at >= seen))
            );
        }
    }
    Ok(json!({"items":items}))
}
